tests/csapi/user_directory_display_names_test.go

```diff
@@ -0,0 +1,64 @@
+// +build !synapse_blacklist
+
+// Rationale for being included in Synapse's blacklist: https://github.com/matrix-org/synapse/issues/5677
+package csapi_tests
+
+import (
+	"testing"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+func TestRoomSpecificUsernameNotLeaked(t *testing.T) {
+	// Reproduces https://github.com/matrix-org/synapse/issues/5677
+	// In that bug report, Alice has revealed a private name to a friend X,
+	// and Bob can see that private name when he shouldn't be able to.
+	// I've tweaked the names to be more traditional: Alice reveals a private name
+	// to Bob, and Eve shouldn't be able to see that name.
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.RegisterUser(t, "hs1", "bob", "bob-pw")
+	eve := deployment.RegisterUser(t, "hs1", "eve", "eve-pw")
+
+	t.Run("Usernames specific to a room aren't leaked in the user directory", func(t *testing.T) {
+		// Bob creates a new room and invites Alice. She accepts.
+		privateRoom := bob.CreateRoom(t, map[string]interface{}{
+			"m.federate": false,
+		})
+		bob.InviteRoom(t, privateRoom, "@alice:hs1")
+		alice.JoinRoom(t, privateRoom, nil)
```

Comment:
This will race on multi-process server implementations (Dendrite, Synapse in worker mode) because the invite may not have fully propagated through the system before the join request is made. You need to wait until you see the invite on alice's sync stream before joining the room.